charts/scalardb-analytics-postgresql/templates/scalardb-analytics-postgresql/configmap.yaml

```diff
@@ -0,0 +1,48 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ include "scalardb-analytics-postgresql.fullname" . }}-database-properties
+  namespace: {{ .Release.Namespace }}
+data:
+  # Create a database.properties file which is config file of ScalarDB Analytics with PostgreSQL.
+  database.properties:
+    {{- toYaml .Values.scalardbAnalyticsPostgreSQL.databaseProperties | nindent 4 }}
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ include "scalardb-analytics-postgresql.fullname" . }}-entrypoint-shell
+  namespace: {{ .Release.Namespace }}
+data:
+  entrypoint.sh: |
+    #!/bin/bash
```

Comment:
At this time, I create the `entorypoint.sh` on the helm chart side. Schema Importer container mounts this file and runs it as an entrypoint. 

The main purpose of this shell is to implement the `retry` process for Schema Importer.

We need this file on the helm chart side to run the existing stable versions (v3.10) images.

However, I will create this `entrypoint.sh` on the Schema Imorter container image side in the future. After that (maybe after v3.11), we can use the `etntrypoint.sh` that is included in the container image.

Reply:
So, this runs Schema Importer every time the pod starts?

Reply:
Yes. That's right. For example, we run Schema Importer in the following cases:

* Deploy pods.
* Pods restart for some reason.
* Scale out the pods.

Reply:
Thank you for the response. I have one more question: why is it necessary to include the Schema Importer in the pod? I'm considering the possibility of running the Schema Importer separately, outside of the pod.

Reply:
@brfrn169  
Thank you for your question!

In conclusion, I want to deploy ScalarDB Analytics with PostgreSQL as a `Stateless` workload to reduce maintenance costs and complex configurations, at the moment. This is why I run the Schema Importer as a sidecar in the pod.

For example, I want to avoid manually running the Schema Importer in the following cases.

* Existing pods crash or restart
* Scale pods (new pods start)

## Challenges

ScalarDB Schema Loader and ScalarDL Schema Loader create database schemas (i.e., create some tables or objects) on the backend database side. In other words, those objects are persisted by the backend database. So, basically, ScalarDB/ScalarDL has no state in themselves.

However, Schema Importer creates some objects (e.g., foreign servers, extensions, and views) on the ScalarDB Analytics with PostgreSQL side. In other words, strictly, ScalarDB Analytics with PostgreSQL has states. It's a `Stateful` workload.

So, if the pod is restarted for some reason, the loaded objects are lost. In this case, we have to re-run Schema Importer to re-load all objects on the PostgreSQL.

As well as the pods restart, we have to run the Schema Importer if we scale out (add a new pod) to create some objects in the new pod (in the new PostgreSQL). This is because each pod (each PostgreSQL) has objects (e.g., views) respectively.

## Solution 1 (make it `Stateful` workload)

To address the above challenges, we can deploy ScalarDB Analytics with PostgreSQL as a `Stateful` workload by using `StatefulSet` which is one of the Kubernetes resources.

In this case, the objects (foreign server, extension, and views) are stored and persisted in the PV (persistent volume) which is attached to the pod.

So, we don't need to re-run Schema Importer if the pods crash/restart. However, we still have to run Schema Importer manually when we scale out pods.

Also, in this solution, we have to use a bit more complex configurations for instance `StatefulSet` and `PersistentVolume`, rather than we deploy it as a `Stateless` workload by using `Deployment`. It takes maintenance costs.

In addition, in this case, we have to consider the backup/restore of the `PersistentVolume`. It increases operation costs.

So, I want to avoid this solution if I can.

## Solution 2 (run Schema Importer manually every time)

This is a simple (but not easy) solution. We can deploy ScalarDB Analytics with PostgreSQL as a `Stateless` workload with `Deployment`, and run Schema Importer manually every time we need.

However, this solution increases the operation costs on the user side. And, this solution cannot take advantage of the self-healing (automatically pod restart when some failure occurs) feature of Kubernetes well.

## Solution 3 (run Schema Importer as a sidecar / our choice)

To resolve the challenges with the smallest additional costs, we decided to run Schema Importer as a sidecar in the pod startup step.

In this case, the pod runs Schema Importer automatically when pods crash/restart or a new pod is added. So, we can avoid manually running Schema Importer.

Also, in this case, we don't need to store/persist some objects in the `PersistentVolume` because the pod runs Schema Importer every time on its startup. We can avoid additional costs to maintain the `Stateful` workload.

This is why I decided to run Schema Importer as a sidecar.

Reply:
Sounds good! Thank you for the explanation!